cmd/api-router-cors.go

```diff
@@ -0,0 +1,234 @@
+//    and was limited to just allowed origins.
+// 2. Bucket-specific CORS settings: This is a new method that allows for more fine-grained
+//    control over CORS settings, and can be applied on a per-bucket basis. It uses AWS's own spec
+//    that allows a number of rules to be specified. Compatible with the global settings, takes precedence
+//    https://docs.aws.amazon.com/AmazonS3/latest/userguide/cors.html
+
+// corsHandler returns a middleware that is able to either apply bucket-specific CORS settings if they exist,
+// or fallback to the global CORS settings if they don't.
+func corsHandler(handler http.Handler) http.Handler {
+	corsGlobal := corsGlobalHandler(handler)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		bucket := vars["bucket"]
+		if bucket == "" {
+			// No bucket, continue the request using global CORS.
+			corsGlobal.ServeHTTP(w, r)
+			return
+		}
+
+		// Check if the bucket has a CORS config set.
+		config, _, err := globalBucketMetadataSys.GetCorsConfig(bucket)
+		if err != nil {
+			if errors.Is(err, BucketCorsNotFound{Bucket: bucket}) {
+				// No CORS config set on the bucket, continue with the request using global CORS.
+				corsGlobal.ServeHTTP(w, r)
+				return
+			}
+
+			if r.Method == http.MethodOptions {
+				// Don't fall back to global CORS on preflight as that presents a security risk.
+				writeErrorResponse(r.Context(), w, toAPIError(r.Context(), err), r.URL)
+				return
+			}
+			// Continue with the non-preflight request
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		// We have bucket specific CORS settings, use them instead of global CORS handling.
+		opts := findCorsRule(r, config)
+		if opts == nil {
+			if r.Method == http.MethodOptions {
+				// Preflight request with no matching rule is an error.
+				writeErrorResponse(r.Context(), w, errorCodes.ToAPIErr(ErrCORSRequestNotAllowed), r.URL)
+				return
+			}
+			// Non-preflight request with no matching rule is allowed.
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		// fetch() spec guarantees all values of Access-Control-Request-Headers are lowercase,
+		// however Go canonicalizes header values to Title-Case. We need to convert them to lowercase
+		// in order to match correctly in rs/cors. Access the map directly, as recommended by http.Header docs.
+		if r.Method == http.MethodOptions {
+			for i, header := range r.Header["Access-Control-Request-Headers"] {
+				r.Header["Access-Control-Request-Headers"][i] = strings.ToLower(header)
+			}
+		}
+
+		// To match S3 behavior, do not send Allow-Credentials header for certain methods.
+		skipCredentialsMethods := []string{http.MethodGet, http.MethodHead}
+		methodToCheck := r.Header.Get("Access-Control-Request-Method")
+		if methodToCheck == "" {
+			methodToCheck = r.Method
+		}
+		if slices.Contains(skipCredentialsMethods, methodToCheck) {
+			opts.AllowCredentials = false
+		}
+
+		rscors.New(*opts).ServeHTTP(w, r, handler.ServeHTTP)
```

Comment:
```suggestion
		rscors.New(*opts).ServeHTTP(w, r, handler.ServeHTTP)
```

is this okay to cause repeated allocation? any overhead?

Reply:
@harshavardhana Yeah this will generate overhead, it's not as efficient as it could be. I would say this PR's approach of using `rs/cors` and the `mux.Middleware` is just the simple thing that works, and has less potential for disruption and maybe less complexity than other alternatives. But it's definitely creating an overhead. 

I have some basic optimizations that I can do for the allocations in `corsGlobalHandler`, I will push up as a new commit.

For this line 107 that you commented on, that's the tricky one to optimize, since it's the case where a CORS Rule (XML rule) has been found for a bucket, and we want to apply it. Going "via" `rs/cors` keeps things consistent, but involves extra allocations. I could remove `rs/cors` itself and then try re-implement the critical parts it in a more efficient way, but not sure how you feel about that.

Reply:
Lets keep this for now, this is an esoteric use-case anyway.